go-controller/pkg/ovn/egressip.go

```diff
@@ -988,29 +988,24 @@ func (oc *Controller) isAnyClusterNodeIP(ip net.IP) *egressNode {
 func (oc *Controller) updateEgressIPStatusWithItem(name string, statusItem egressipv1.EgressIPStatusItem) error {
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		// Get the latest item from the store before updating.
-		egressIP, err := oc.watchFactory.GetEgressIP(name)
+		egressIP, err := oc.kube.GetEgressIP(name)
```

Comment:
This doesn't really have anything to do with guarding against races in our own code. Keep in mind that this is already done by means of the `egressIPAssignmentMutex`. The reason for this change is to make sure we don't act on stale data when we need to perform multiple updates in a row. Consider the following:

You have an egress IP object with status: [1, 2], you want to add 3 and 4 sequentially using this function. If we are using the informer cache, so ending up in the following situation takes little to no effort:

1. `updateEgressIPStatusWithItem` is called, adding 3 to [1,2]. The informer cache is up-to-date so the object retrieved does have status [1,2]
2. `updateEgressIPStatusWithItem` is called again, adding 4 to [1,2,3]. The informer cache is not up-to-date because it still hasn't been updated by the update made in 1., so the object returned has status [1,2]
3. The update in 1. is overwritten by the one made in 2. leading to a final state of: [1,2,4]

You can't rely on the informer cache when you are both reading and writing to the same object and field. You can when only reading and you have your own separate data store to update/write to.

Reply:
Just re-iterating our offline discussion here. Alex is going to look into using Patch operation instead of Update to see if that will be good enough for adding/removing EgressIPStatusItems.

Reply:
Using a patch operation saves us from retrieving the object just before we update it, which we did originally because we needed to be sure that we didn't overwrite any user changes for updates which might take a long time. Given that patch operations can specify which individual field to update, we don't have this problem anymore and do not need to retrieve the object just before updating it. 

As to removing all client calls used by the egress IP code, that can't be solved by this. Patch operations are limited by  [RFE 6902](https://datatracker.ietf.org/doc/html/rfc6902#appendix-A.4) which only allows removal of slice elements by the index, hence we can't really use it since that model isn't "concurrency-safe". An example of this is: say we have a status [1,2,3] and we want to remove [2,3]. To do that we'd need to remove index 1 first and then 2, but by the time 1 has been removed, 2 is not at index 2 anymore. 

We still need the `WatchEgressNodes` and `reconcileCloudPrivateIPConfig` functions to retrieve the last state of the egress IP object and pass that to the `reconcileEgressIP` so that we  are sure to always update the latest version.